### ISSUE TITLE
No Memory errors on Macs

### ISSUE THREAD

**Issue by andymeneely** [andymeneely/squib#88]
Cairo seems to think that it's out of memory on these machines:

http://www.boardgamegeek.com/article/18697046#18697046
http://www.boardgamegeek.com/article/19660991#19660991

```
/Users/andymatthews/.rvm/gems/ruby-2.0.0-p643/gems/squib-0.6.0/lib/squib/card.rb:52:in `use_cairo': out of memory (NoMemoryError)
from /Users/andymatthews/.rvm/gems/ruby-2.0.0-p643/gems/squib-0.6.0/lib/squib/graphics/text.rb:165:in `text'
from /Users/andymatthews/.rvm/gems/ruby-2.0.0-p643/gems/squib-0.6.0/lib/squib/api/text.rb:54:in `block in text'
from /Users/andymatthews/.rvm/gems/ruby-2.0.0-p643/gems/squib-0.6.0/lib/squib/api/text.rb:53:in `each'
from /Users/andymatthews/.rvm/gems/ruby-2.0.0-p643/gems/squib-0.6.0/lib/squib/api/text.rb:53:in `text'
from deck.rb:4:in `block in <main>'
from /Users/andymatthews/.rvm/gems/ruby-2.0.0-p643/gems/squib-0.6.0/lib/squib/deck.rb:71:in `instance_eval'
from /Users/andymatthews/.rvm/gems/ruby-2.0.0-p643/gems/squib-0.6.0/lib/squib/deck.rb:71:in `initialize'
from deck.rb:3:in `new'
from deck.rb:3:in `<main>'
```

Environment:
```
rvm 1.26.11
ruby 2.0.0p643
OSX 10.10.3
```

They appear to be from the hello world sample, so it's clearly not out of memory.

**Comment by andymeneely** [comment]
If you are having this problem, can you please run this on your machine? I'm curious if you get the memory error with this.

```ruby
require 'cairo'
require 'pango'

cxt = Cairo::Context.new(Cairo::ImageSurface.new(100, 100))
cxt.save
layout = cxt.create_pango_layout
layout.text = 'Hello, world!'
cxt.show_pango_layout(layout)
cxt.restore
cxt.target.write_to_png('hello_world.png')
```

**Comment by ckolderup** [comment]
Hi,

I was getting the out of memory errors trying to load just the sample code deck ("Hello" and "World" cards), so I tried the above.

I get the following error running that ruby file:

```
$ bundle exec ruby cairo.rb
cairo.rb:9:in `restore': out of memory (NoMemoryError)
	from cairo.rb:9:in `<main>'
```

I'm running the following versions of things:
* Mac OS X Yosemite 10.10.4 on a 13" Retina MBP with 8GB of RAM and over 5GB free right now
* homebrew 0.9.5
* cairo: stable 1.14.2
* rbenv 0.4.0
* ruby: 2.2.1 (via rbenv)
* squib 0.6.0 via bundler with these resolved gem dependencies in Gemfile.lock:
```
squib (0.6.0)
  cairo (~> 1.14.0)
  mercenary (~> 0.3.4)
  pango (~> 2.2.0)
  roo (~> 2.0.0)
  rsvg2 (~> 2.2.0)
  ruby-progressbar (~> 1.6)
```

Pretty new to Cairo and all the graphics stuff so I can try poking at it but let me know if there's more directed work I can do to help. Thanks!

**Comment by andymeneely** [comment]
Thanks @ckolderup! 

Ok, so it looks like it is a Cairo/Pango thing and not a Squib thing. I'm thinking at this point we report it to the rcairo project, but it might get passed up to the native Cairo eventually. I have seen several messages recently about Mac support in the Cairo mailing list (mostly related to supporting really old OSX releases, not memory management). I'm thinking it's a compilation issue where malloc() is not behaving in the way rcairo expects it to.

Also, @ckolderup  - did it show the full backtrace? I'm curious if it failed on `restore` as it has for other folks.

Or it could be pango-related. What's your pango version? Not just the gem, the one that brew tells you. 

Let's try the Cairo hello world without Pango:

```ruby
require 'cairo'

cxt = Cairo::Context.new(Cairo::ImageSurface.new(100, 100))
cxt.save
cxt.set_source_color('blue')
cxt.paint
cxt.restore
cxt.target.write_to_png('hello_world.png')
```
Should produce a blue image. 

Either way, I'll ask you to report this issue to https://github.com/rcairo/rcairo. Be sure to provide the snippet.

By the way, for me, I've got Yosemite 10.10.2, and running with Cairo 1.14.2 (the rcairo gem is at 1.14.1 which is the latest currently). I'm using rvm instead of rbenv, but with Ruby 2.2.1p85 these scripts work for me. My Pango gem is at 2.2.5 and the native library is at 1.36.8.

It's strange that this is happening with Ruby 2.2 - in the past it's always been Ruby 2.1. Are you sure that you're on Ruby 2.2.1? Do `ruby --version` just for my sanity.

**Comment by mark** [comment]
I'm having the same issue... your Cairo hello world is working fine.  brew info pango tells me it's 1.36.8, and the ruby gem is 2.2.5.

**Comment by andymeneely** [comment]
Ok, so it's pango. The only difference I see is Yosemite 10.10.2 vs 10.10.4 - right? 

Would someone be willing to submit this to the ruby-gnome2 Pango page? Link is below. Be sure to mention that, even though it's a Cairo error, it only fails when we bring in Pango. And it's on the "restore".

https://github.com/ruby-gnome2/ruby-gnome2/tree/master/pango

**Comment by mark** [comment]
I created the following issue: https://github.com/ruby-gnome2/ruby-gnome2/issues/477

If anyone following this has more info, please add it there.  Hopefully someone familar with pango knows what's going on and how to fix it.  Looking for similar problems, I found suggestions that it might be due to Cairo not being able to find fonts, or cairo (or related libraries) being compiled with clang instead of gcc, but I couldn't find anything conclusive.  Or that worked at fixing the issue.

**Comment by mark** [comment]
@andymeneely what do you see when you run `GCC --version` on the command line?

**Comment by andymeneely** [comment]
Here's my gcc info:

```
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 5.1 (clang-503.0.40) (based on LLVM 3.4svn)
Target: x86_64-apple-darwin14.1.0
Thread model: posix
```

I don't think it would get there if it had trouble finding fonts. I've run into that before in trying to get Squib to run in Cygwin, and the errors came earlier than save/restore.

**Comment by mark** [comment]
This is what I see... not sure if it indicated anything or not:

```
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 6.1.0 (clang-602.0.53) (based on LLVM 3.6.0svn)
Target: x86_64-apple-darwin14.4.0
Thread model: posix
```

**Comment by mostlybadfly** [comment]
Not sure if this helps, but I had the same problem when I was still on ruby 2.1, it seems to work just fine after switching to 2.2.1
```
magicant:my-cool-game mostlybadfly$ rvm use 2.1
Using /Users/mostlybadfly/.rvm/gems/ruby-2.1.5
magicant:my-cool-game mostlybadfly$ ruby deck.rb
/Users/mostlybadfly/.rvm/gems/ruby-2.1.5/gems/squib-0.6.0/lib/squib/card.rb:52:in `use_cairo': out of memory (NoMemoryError)
	from /Users/mostlybadfly/.rvm/gems/ruby-2.1.5/gems/squib-0.6.0/lib/squib/graphics/text.rb:165:in `text'
	from /Users/mostlybadfly/.rvm/gems/ruby-2.1.5/gems/squib-0.6.0/lib/squib/api/text.rb:54:in `block in text'
	from /Users/mostlybadfly/.rvm/gems/ruby-2.1.5/gems/squib-0.6.0/lib/squib/api/text.rb:53:in `each'
	from /Users/mostlybadfly/.rvm/gems/ruby-2.1.5/gems/squib-0.6.0/lib/squib/api/text.rb:53:in `text'
	from deck.rb:4:in `block in <main>'
	from /Users/mostlybadfly/.rvm/gems/ruby-2.1.5/gems/squib-0.6.0/lib/squib/deck.rb:71:in `instance_eval'
	from /Users/mostlybadfly/.rvm/gems/ruby-2.1.5/gems/squib-0.6.0/lib/squib/deck.rb:71:in `initialize'
	from deck.rb:3:in `new'
	from deck.rb:3:in `<main>'
magicant:my-cool-game mostlybadfly$ rvm use 2.2
Using /Users/mostlybadfly/.rvm/gems/ruby-2.2.1
magicant:my-cool-game mostlybadfly$ ruby deck.rb
```

**Comment by commadelimited** [comment]
I'm having the same results as @ckolderup and others.

`test-cairo-pango.rb` fails with:

```
☀  my-cool-game  touch test-cairo-pango.rb
☀  my-cool-game  ruby test-cairo-pango.rb
test-cairo-pango.rb:9:in `restore': out of memory (NoMemoryError)
	from test-cairo-pango.rb:9:in `<main>'
```
while `test-cairo.rb` produces a blue image as expected:

```
☀  my-cool-game  touch test-cairo.rb
☀  my-cool-game  ruby test-cairo.rb
```

My GCC:
```
☀  my-cool-game  GCC --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 6.1.0 (clang-602.0.53) (based on LLVM 3.6.0svn)
Target: x86_64-apple-darwin14.3.0
Thread model: posix
```

I can confirm though that after changing Ruby to 2.2.1 as reported by @mostlybadfly, I can successfully run the original sample app.

**Comment by andymeneely** [comment]
So, to sum up:
  * Ruby >= 2.2 seems to work for most people, except @ckolderup.
  * Both LLVM 6.1.0 and LLVM 5.1 have worked when we upgrade to 2.2. 
  * Pango is still related to all this as well, although it's a Cairo error on `restore`

@ckolderup - are you positive you're on Ruby 2.2? I'm not familiar with rbenv beyond the basics - I usually stick with rvm.

**Comment by ckolderup** [comment]
@andymeneely yep, I know for sure I'm on 2.2.1 right now. I'm going to uninstall it, upgrade rbenv, and reinstall the most recent 2.2 build they can give me and see if that changes anything—I'll report back shortly.

**Comment by ckolderup** [comment]
Upgrading everything including a fresh build of ruby via rbenv/`ruby-build` fixed it! If it helps anyone else, here's a revised list of my versions of everything:

Mac OS X Yosemite 10.10.4
homebrew 0.9.5
cairo: stable 1.14.2
rbenv 0.4.0
ruby: 2.2.2 (via rbenv)
squib: 0.6.0
cairo: 1.14.1
pango: 2.2.5

Thanks for the help!

**Comment by andymeneely** [comment]
Ok then. Looks like a Ruby upgrade is all we need. I know that Ruby 2.2 did an overhaul on the garbage collector, so it could be related to that. I'm closing this one for now, but I'm willing to have it reopened if it rears it's ugly head. 

Thanks everyone!

TL;DR for future people looking into this: the fix is to get Ruby 2.2.
